File: backend/app/api/logs.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required
from app.middleware.rbac import admin_required
from app.services.log_service import LogService

logs_bp = Blueprint('logs', __name__)


@logs_bp.route('', methods=['GET'])
@jwt_required()
@admin_required
def list_log_files():
    """List available log files."""
    logs = LogService.get_log_files()
    return jsonify({'logs': logs}), 200


@logs_bp.route('/read', methods=['GET'])
@jwt_required()
@admin_required
def read_log():
    """Read lines from a log file."""
    filepath = request.args.get('path')
    lines = request.args.get('lines', 100, type=int)
    from_end = request.args.get('from_end', 'true').lower() == 'true'

    if not filepath:
        return jsonify({'error': 'path parameter is required'}), 400

    result = LogService.read_log(filepath, lines, from_end)
    return jsonify(result), 200 if result['success'] else 400


@logs_bp.route('/search', methods=['GET'])
@jwt_required()
@admin_required
def search_log():
    """Search log file for a pattern."""
    filepath = request.args.get('path')
    pattern = request.args.get('pattern')
    lines = request.args.get('lines', 100, type=int)

    if not filepath or not pattern:
        return jsonify({'error': 'path and pattern parameters are required'}), 400

    result = LogService.search_log(filepath, pattern, lines)
    return jsonify(result), 200 if result['success'] else 400


@logs_bp.route('/app/<app_name>', methods=['GET'])
@jwt_required()
@admin_required
def get_app_logs(app_name):
    """Get logs for a specific application."""
    log_type = request.args.get('type', 'access')
    lines = request.args.get('lines', 100, type=int)

    result = LogService.get_app_logs(app_name, log_type, lines)
    return jsonify(result), 200 if result['success'] else 400


@logs_bp.route('/journal', methods=['GET'])
@jwt_required()
@admin_required
def get_journal_logs():
    """Get logs from systemd journal."""
    unit = request.args.get('unit')
    lines = request.args.get('lines', 100, type=int)
    since = request.args.get('since')
    priority = request.args.get('priority')

    result = LogService.get_journalctl_logs(unit, lines, since, priority)
    return jsonify(result), 200 if result['success'] else 400


@logs_bp.route('/clear', methods=['POST'])
@jwt_required()
@admin_required
def clear_log():
    """Clear/truncate a log file."""
    data = request.get_json()
    filepath = data.get('path') if data else None

    if not filepath:
        return jsonify({'error': 'path is required'}), 400

    result = LogService.clear_log(filepath)
    return jsonify(result), 200 if result['success'] else 400


@logs_bp.route('/rotate', methods=['POST'])
@jwt_required()
@admin_required
def rotate_logs():
    """Trigger log rotation."""
    result = LogService.rotate_logs()
    return jsonify(result), 200 if result['success'] else 400

File: backend-rs/crates/sk-ops/src/logs.rs
//! `LogService` port: log file listing, tail/head reads, grep search,
//! journald, truncation. Same path-allowlist protection as Flask.

use crate::{run, sourced_result};
use serde_json::{json, Value};

/// `LogService.LOG_PATHS`
const LOG_PATHS: &[(&str, &str)] = &[
    ("nginx_access", "/var/log/nginx/access.log"),
    ("nginx_error", "/var/log/nginx/error.log"),
    ("php_fpm", "/var/log/php*-fpm.log"),
    ("mysql", "/var/log/mysql/error.log"),
    ("postgresql", "/var/log/postgresql/postgresql-*-main.log"),
    ("syslog", "/var/log/syslog"),
    ("auth", "/var/log/auth.log"),
];

/// `LogService.ALLOWED_LOG_DIRECTORIES` (SERVERKIT_DIR ~ /opt/serverkit is
/// covered by /opt).
const ALLOWED_DIRS: &[&str] = &["/var/log", "/var/www", "/home", "/opt"];

/// Path-traversal protection: resolve symlinks, then prefix-match.
pub fn is_path_allowed(filepath: &str) -> bool {
    match std::fs::canonicalize(filepath) {
        Ok(real) => {
            let real = real.to_string_lossy();
            ALLOWED_DIRS.iter().any(|d| real.starts_with(d))
        }
        Err(_) => false,
    }
}

fn format_size(bytes: u64) -> String {
    let mut val = bytes as f64;
    for unit in ["B", "KB", "MB", "GB"] {
        if val < 1024.0 {
            return format!("{val:.1} {unit}");
        }
        val /= 1024.0;
    }
    format!("{val:.1} TB")
}

/// `LogService.get_log_files()`
pub fn log_files() -> Vec<Value> {
    let mut logs = Vec::new();
    for (name, pattern) in LOG_PATHS {
        let Ok(paths) = glob::glob(pattern) else {
            continue;
        };
        for path in paths.flatten() {
            let Ok(meta) = std::fs::metadata(&path) else {
                continue;
            };
            let modified = meta
                .modified()
                .ok()
                .map(|t| {
                    chrono::DateTime::<chrono::Local>::from(t)
                        .naive_local()
                        .format("%Y-%m-%dT%H:%M:%S%.6f")
                        .to_string()
                })
                .unwrap_or_default();
            logs.push(json!({
                "name": name,
                "path": path.to_string_lossy(),
                "size": meta.len(),
                "size_human": format_size(meta.len()),
                "modified": modified,
            }));
        }
    }
    logs
}

/// `LogService.read_log()` — tail/head with Rust-I/O fallback.
pub async fn read_log(filepath: &str, lines: i64, from_end: bool) -> Value {
    if !is_path_allowed(filepath) {
        return json!({ "success": false, "error": "Access denied: path not in allowed directories" });
    }
    if !std::path::Path::new(filepath).exists() {
        return json!({ "success": false, "error": "Log file not found" });
    }

    let tool = if from_end { "tail" } else { "head" };
    match run(tool, &["-n", &lines.to_string(), filepath], 30).await {
        Ok((true, stdout, _)) => {
            let log_lines: Vec<String> = stdout.split('\n').map(str::to_string).collect();
            let mut result = sourced_result(log_lines, tool, tool);
            result["filepath"] = json!(filepath);
            result
        }
        Ok((false, _, stderr)) => json!({ "success": false, "error": stderr }),
        Err(_) => {
            // Fallback: direct read (tail/head unavailable)
            match std::fs::read_to_string(filepath) {
                Ok(content) => {
                    let all: Vec<&str> = content.lines().collect();
                    let n = lines.max(0) as usize;
                    let slice: Vec<String> = if from_end {
                        all.iter()
                            .rev()
                            .take(n)
                            .rev()
                            .map(|s| s.to_string())
                            .collect()
                    } else {
                        all.iter().take(n).map(|s| s.to_string()).collect()
                    };
                    let mut result = sourced_result(slice, "rust", "direct file read");
                    result["filepath"] = json!(filepath);
                    result
                }
                Err(e) => json!({ "success": false, "error": e.to_string() }),
            }
        }
    }
}

/// `LogService.search_log()` — case-insensitive grep, max N matches.
pub async fn search_log(filepath: &str, pattern: &str, lines: i64) -> Value {
    if !is_path_allowed(filepath) {
        return json!({ "success": false, "error": "Access denied: path not in allowed directories" });
    }
    if !std::path::Path::new(filepath).exists() {
        return json!({ "success": false, "error": "Log file not found" });
    }

    match run(
        "grep",
        &["-i", "-m", &lines.to_string(), pattern, filepath],
        60,
    )
    .await
    {
        // grep exit 1 = no matches, not an error
        Ok((ok, stdout, stderr)) => {
            if ok || stdout.is_empty() && stderr.is_empty() {
                let matches: Vec<&str> = stdout.lines().filter(|l| !l.is_empty()).collect();
                json!({
                    "success": true,
                    "matches": matches,
                    "count": matches.len(),
                    "pattern": pattern,
                })
            } else {
                json!({ "success": false, "error": stderr })
            }
        }
        Err(e) => json!({ "success": false, "error": e }),
    }
}

/// `LogService.get_journalctl_logs()` (journalctl source; syslog fallback).
pub async fn journal_logs(
    unit: Option<&str>,
    lines: i64,
    since: Option<&str>,
    priority: Option<&str>,
) -> Value {
    let lines_s = lines.to_string();
    let mut args: Vec<&str> = vec!["-n", &lines_s, "--no-pager", "-o", "short-iso"];
    if let Some(u) = unit {
        args.extend(["-u", u]);
    }
    if let Some(s) = since {
        args.extend(["--since", s]);
    }
    if let Some(p) = priority {
        args.extend(["-p", p]);
    }

    match run("journalctl", &args, 60).await {
        Ok((true, stdout, _)) => {
            let log_lines: Vec<String> = stdout.split('\n').map(str::to_string).collect();
            sourced_result(log_lines, "journalctl", "systemd journal")
        }
        Ok((false, _, stderr)) => json!({ "success": false, "error": stderr }),
        Err(_) => {
            // Syslog fallback
            read_log("/var/log/syslog", lines, true).await
        }
    }
}

/// `LogService.clear_log()` — truncate to zero.
pub async fn clear_log(filepath: &str) -> Value {
    if !is_path_allowed(filepath) {
        return json!({ "success": false, "error": "Access denied: path not in allowed directories" });
    }
    if !std::path::Path::new(filepath).exists() {
        return json!({ "success": false, "error": "Log file not found" });
    }
    match run("truncate", &["-s", "0", filepath], 15).await {
        Ok((true, _, _)) => {
            json!({ "success": true, "message": format!("Log file {filepath} cleared") })
        }
        Ok((false, _, stderr)) => json!({ "success": false, "error": stderr }),
        Err(e) => json!({ "success": false, "error": e }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn allowlist_blocks_traversal() {
        assert!(!is_path_allowed("/etc/shadow"));
        assert!(!is_path_allowed("/var/log/../../etc/passwd"));
    }
}
